src/odemis/gui/cont/tabs.py

```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+                                align_mode)
                 return
 
-            # For streak-focus, it's not useful to use turnLightOn(), as the camera should not
-            # yet be receiving light anyway.
-            # TODO: if standard focus mode, call Sparc2AutoFocus(align_mode, main.opm, s, start_autofocus=False)
-
-            # Note: First close slit, then switch on calibration light
-            # Go to the special focus mode (=> close the slit)
-            f = main.opm.setPath(opath).result()
-
-            # force wavelength 0
-            main.spectrograph.moveAbsSync({"wavelength": 0})
-
             if align_mode != "streak-align":
                 self._stream_controller.pauseStreams()
 
-            # Turn on the light
-            bl.power.value = bl.power.range[1]
-            if s:
-                s.should_update.value = True  # the stream will set all the emissions to 1
-            else:
-                bl.emissions.value = [1] * len(bl.emissions.value)
-
-        else:  # manual focus button is inactive
-            # Go back to previous mode (=> open the slit & turn off the lamp)
             if align_mode == "streak-align":
-                s = None
-            # Note: First turn off calibration light, then open slit
-            bl.power.value = bl.power.range[0]
-            if s:
-                s.should_update.value = True  # the stream will set all the emissions to 0
-            else:
-                bl.emissions.value = [0] * len(bl.emissions.value)
+                # force wavelength 0
+                # TODO: Make sure it's the correct position on the workflow, maybe do it for all modes?
+                main.spectrograph.moveAbsSync({"wavelength": 0})
 
+            self._mf_future = Sparc2ManualFocus(main.opm, bl, opath, mf_toggled=True)
+            self._mf_future.add_done_callback(self._onManualFocusReady)
+            # Update GUI
+            self._pfc_manual_focus = ProgressiveFutureConnector(self._mf_future, gauge)
+        else:  # manual focus button is untoggled
+            # Go back to previous mode (=> open the slit & turn off the lamp)
+            # Get the optical path from align mode
             opath = self._mode_to_opm[align_mode]
-            f = main.opm.setPath(opath)
-            f.add_done_callback(self._on_align_mode_done)
-            f.result()  # TODO: don't block the GUI => make it part of the manual focus??
-            # if align_mode != "streak-align":
-            #     # Enable autofocus if manual focus btn not pushed
-            #     self.panel.btn_autofocus.Enable(True)
-            #     self.panel.gauge_autofocus.Enable(True)
 
+            self._mf_future = Sparc2ManualFocus(main.opm, bl, opath, mf_toggled=False)
+            self._mf_future.add_done_callback(self._onManualFocusFinished)
+            # Update GUI
+            self._pfc_manual_focus = ProgressiveFutureConnector(self._mf_future, gauge)
+
+    def _onManualFocusReady(self, future):
```

Comment:
You should protect this function with a `@call_in_wx_main`, as it will be run in a separate thread and it accesses the GUI.